lisp/ess-developer.el

```diff
@@ -30,130 +30,248 @@
+    (setq-local ess-developer-local-package pkg-info)))
+
+(defun ess-developer-inject-to-package (&optional attached-only)
+  "Select a package for ESS developer functions and add it
+permanently to the file-local variables section. If ATTACHED-ONLY
+is non-nil, only prompt for attached packages.
+
+See also `ess-developer-inject-to-current-env'."
   (interactive)
-  (unless ess-developer-packages
-    (error "Nothing to remove, 'ess-developer-packages' is empty"))
-  (let ((sel (ess-completing-read "Remove package(s)"
-                                  (append ess-developer-packages (list "*ALL*"))
-                                  nil t nil nil "*ALL*")))
-    (if (equal "*ALL*" sel)
-        (progn
-          (setq ess-developer-packages nil)
-          (ess-developer-deactivate-in-package nil 'all)
-          (message "Removed *ALL* packages from the `ess-developer-packages' list."))
-      (setq ess-developer-packages (delete sel ess-developer-packages))
-      (ess-developer-deactivate-in-package sel 'all)
-      (message "Removed package '%s' from the `ess-developer-packages' list"
-               (propertize sel 'face 'font-lock-function-name-face)))))
+  (let ((pkg-name (ess-developer--select-package-name attached-only)))
+    (setq-local ess-developer-code-injection pkg-name)
+    (message (format "Injecting code in %s" pkg-name))))
+
+(defun ess-developer-activate-code-injection (&optional attached-only)
+  "Select current environment for code injection. Useful when
+debugging a function.
+
+See also `ess-developer-inject-to-package'."
+  (interactive)
+  (cond ((string= ess-developer-code-injection "*current*")
+         (ess-developer-inject-to-package attached-only))
+        (t
+         (setq-local ess-developer-code-injection "*current*")
+         (message "Injecting code in *current*"))))
+
+(defun ess-developer-activate-injection-in-package ()
+  (when ess-developer-code-injection-in-packages
+    (setq-local ess-developer-code-injection
+                (car (ess-developer-current-package-info)))))
+
+(add-hook 'R-mode-hook 'ess-developer-activate-injection-in-package)
+
+(defun ess-developer-send-process (command &optional msg alt)
+  (ess-force-buffer-current)
+  (let* ((pkg-info (or (ess-developer-current-package-info)
+                       (ess-developer-select-package)))
+         (name (car pkg-info))
+         (path (concat "'" (cdr pkg-info) "'"))
+         (alt (cond ((stringp alt) alt)
+                    (alt "")))
+         (args (when alt
+                 (read-string "Arguments: " alt)))
+         (args (unless (or (null args)
+                           (string= "" args))
+                 (concat ", " args))))
+    (message msg name)
+    (ess-developer--update-process-local-pkg pkg-info)
+    (ess-eval-linewise (format command (concat path args)))))
+
+
+;;; Package Detection
+
+(defun ess-developer--find-package-info ()
+  "Parses DESCRIPTION file in PATH (R specific so far). PATH
```

Comment:
First line of doc strings must be self contained sentence.  See [here](http://www.gnu.org/software/emacs/manual/html_node/elisp/Documentation-Tips.html) for conventions. 

Please run M-x checkdoc on new modules and docstrings that you add. Our older modules are screwed, but at least we should keep good doc strings for new stuff.